### DISCUSSION TITLE
Visualise code execution

### DISCUSSION THREAD

**Discussion by OnnoEbbens** [hedyorg/hedy#952] [Implementation discussions]
During Python courses I always use https://pythontutor.com/ to visualize how code is executed. On this website you can:
- run your code line by line.
- see the variables and their values as their are created in the memory.
- see the output of the code

This gives valuable insight in what is going on when the computer evaluates code. I think it would be nice to have a simplified version of this in Hedy. How exactly I am not sure yet.

I was thinking about a special tab where people can visualize their code to see what is happening but don't have the option to save or share code. We could also make this a teacher only option so teachers can show this during an explanation of what a variable is. I do not think we should add this as an option to every code frame because it might be confusing. In the end you don't want to constantly visualize your variables in memory.

Maybe this is more of a discussion than an issue please let me know if this should be moved to the discussion section.

**Comment by Felienne** [comment]
This is a really lovely idea! It could be a student project maybe!

The generic version (with visuals) is really hard, but a simple first version could be to just executed smaller versions of a program, first line 1, then line 1 and 2 etc, with a special button (in a special view maybe).
